### PR TITLE
Add stateful testing against numpy arrays

### PR DESCRIPTION
**Pull request by dcherian**
Here is some testing technology I'm in love with nowadays. I am optimistic it can help build some confidence in the array expressions work. The intent here is to test correctness, not the optimizations (it would be fun to figure out property testing for the optimization rules)

It generates a random sequence of operations, applies them to both dask and numpy arrays; and then compares them.

Run it with
```
python -m pytest dask/tests/test_stateful_array.py --hypothesis-verbosity=verbose --array-expr -s
```

 to see a lot of informative output.

you'll need the hypothesis package installed.

It seems to be finding some edge-case bugs. The most severe one I've seen so far is a hang in `broadcast`. I'll add instructions to reproduce in a comment.

TODO:
- [ ] add binary ops against `self.numpy_array` or `self.dask_array`.
- [ ] vindex
- [ ] oindex
- [ ] shuffle / take

### CODE REVIEW

File: dask/tests/test_stateful_array.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        """Apply basic indexing to both arrays."""
+        # Generate a valid basic index for the current shape
+        idx = data.draw(npst.basic_indices(shape=self.shape))
+        note(
+            f"Basic indexing: {idx}, shape {self.shape} -> {self.numpy_array[idx].shape}"
+        )
+        self.numpy_array = self.numpy_array[idx]
+        self.dask_array = self.dask_array[idx]
+
+    @precondition(lambda self: False)
```
Comment (dcherian) [inline dask/tests/test_stateful_array.py:162]: remove this precondition to reproduce a hang with broadcast with `--hypothesis-seed=307603228577777922830132553021337866664` . YMMV with the seed, but the hang should reproduce even without that.

**Comment by mrocklin** [comment]
Oh fun.  I'm excited to work with this (I'm pretty curious about workflows that might help AI be more helpful in developing software, and this seems potentially useful).  

Question though, I'm getting *a lot* of output, only a small amount of which I suspect are genuine issues.  Is there a good way to filter things here to genuine failures?  I'd love something that generated fail cases to feed into CC.

**Comment by mrocklin** [comment]
Ah, I'm only just now noticing the --verbose and -s flags.  Working through this...

**Comment by mrocklin** [comment]
Yeah, this is cool.  Thanks @dcherian .  I'm looking forward to playing with this.

**Comment by dcherian** [comment]
Nice, glad you got oriented.

It is random search through what  could be a very large parameter space. So I'm thinking we probably want one `rule` per "layer" .

PS: One blogpost I've found useful is https://hypothesis.works/articles/how-not-to-die-hard-with-hypothesis/
